core/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/HistogramParser.java

```diff
@@ -46,93 +44,51 @@
+
+        parser.declareField(HistogramAggregationBuilder::order, HistogramParser::parseOrder,
+                Histogram.ORDER_FIELD, ObjectParser.ValueType.OBJECT);
     }
 
     @Override
-    protected boolean token(String aggregationName, String currentFieldName, Token token,
-                            XContentParseContext context, Map<ParseField, Object> otherOptions) throws IOException {
-        XContentParser parser = context.getParser();
-        if (token.isValue()) {
-            if (context.matchField(currentFieldName, Histogram.INTERVAL_FIELD)) {
-                otherOptions.put(Histogram.INTERVAL_FIELD, parser.doubleValue());
-                return true;
-            } else if (context.matchField(currentFieldName, Histogram.MIN_DOC_COUNT_FIELD)) {
-                otherOptions.put(Histogram.MIN_DOC_COUNT_FIELD, parser.longValue());
-                return true;
-            } else if (context.matchField(currentFieldName, Histogram.KEYED_FIELD)) {
-                otherOptions.put(Histogram.KEYED_FIELD, parser.booleanValue());
-                return true;
-            } else if (context.matchField(currentFieldName, Histogram.OFFSET_FIELD)) {
-                otherOptions.put(Histogram.OFFSET_FIELD, parser.doubleValue());
-                return true;
-            } else {
-                return false;
-            }
-        } else if (token == XContentParser.Token.START_OBJECT) {
-            if (context.matchField(currentFieldName, Histogram.ORDER_FIELD)) {
-                InternalOrder order = null;
-                while ((token = parser.nextToken()) != XContentParser.Token.END_OBJECT) {
-                    if (token == XContentParser.Token.FIELD_NAME) {
-                        currentFieldName = parser.currentName();
-                    } else if (token == XContentParser.Token.VALUE_STRING) {
-                        String dir = parser.text();
-                        boolean asc = "asc".equals(dir);
-                        if (!asc && !"desc".equals(dir)) {
-                            throw new ParsingException(parser.getTokenLocation(), "Unknown order direction in aggregation ["
-                                    + aggregationName + "]: [" + dir
-                                    + "]. Should be either [asc] or [desc]");
-                        }
-                        order = resolveOrder(currentFieldName, asc);
-                    }
+    public AggregationBuilder parse(String aggregationName, QueryParseContext context) throws IOException {
+        return parser.parse(context.parser(), new HistogramAggregationBuilder(aggregationName), context);
+    }
+
+    private static InternalOrder parseOrder(XContentParser parser, QueryParseContext context) throws IOException {
+        InternalOrder order = null;
+        Token token;
+        String currentFieldName = null;
+        while ((token = parser.nextToken()) != XContentParser.Token.END_OBJECT) {
```

Comment:
This looks fairly similar to the one for dates above. I wonder if they should share code?

I have this problem - when it looks like code is the same at first glance I have a *ton* of trouble spotting small differences, especially in code reviews. Which is why I'm more sensitive than most folks to this topic.